.github/workflows/ci.yml

```diff
@@ -83,37 +83,47 @@ jobs:
           ref: ${{ github.event.pull_request.head.sha }} # checkout PR HEAD commit instead of merge commit
-      - uses: ./.github/actions/setup
       - name: Check Commits
         uses: trinodb/github-actions/block-commits@c2991972560c5219d9ae5fb68c0c9d687ffcdd10
         with:
           action-merge: fail
           action-fixup: none
-      - name: Check if all intermediate commits of this PR compile
-        run: |
-          set -x
-          # Fake identity to tag rebased commits with.
-          # These commits will be discarded, but the important part is that `git rebase` executes commands between all commits.
-          git config user.name "Compile all commits builder"
-          git config user.email "compile-all-commits@trino.io"
-          # Show the entire PR branch and the base ref all the way to the fork point.
-          # This shows some context on the state of the repo and shows commit hashes without having to scroll through entire repo history.
-          git log --oneline --decorate=full --graph "$( git merge-base HEAD refs/remotes/origin/${{ github.event.pull_request.base.ref }} )~..refs/remotes/origin/${{ github.event.pull_request.base.ref }}" HEAD
-          # We can skip compiling the top commit since it's already being compiled and tested in other jobs.
-          git checkout HEAD~
-          # If the PR branch contains only 1 commit, checkout will move HEAD to a commit already on the base branch, and the rebase doesn't do anything.
-          export MAVEN_OPTS="${MAVEN_INSTALL_OPTS}"
-          git rebase --verbose --exec "$MAVEN package ${MAVEN_COMPILE_COMMITS} ${MAVEN_GIB} || (git status && false)" $( git merge-base HEAD refs/remotes/origin/${{ github.event.pull_request.base.ref }} )
-      # The local repo is broken at this point, and should be abandoned
-      - name: Clean local Maven repo
-        # Avoid creating a cache entry because this job doesn't download all dependencies
-        if: steps.cache.outputs.cache-hit != 'true'
-        run: rm -rf ~/.m2/repository
+      - name: Set matrix
+        id: set-matrix
+        # The output from rev-list ends with a newline, so we have to filter out index -1 in jq since it's an empty string
+        # The HEAD commit of the PR (index -2) can be safely ignored since it's already compiled in other jobs
+        run: |
+          export JQ_PIPELINE='split("\n") | .[0:-2] | map({base_ref: "${{ github.event.pull_request.base.ref }}", commit: .}) | {include: .}'
+          git rev-list refs/remotes/origin/${{ github.event.pull_request.base.ref }}..HEAD | jq --raw-input --slurp "$JQ_PIPELINE" > commit-matrix.json
+          echo "Commit matrix: $(jq '.' commit-matrix.json)"
+          echo "matrix=$(jq -c '.' commit-matrix.json)" >> $GITHUB_OUTPUT
+
+  check-commits-dispatcher:
+    runs-on: ubuntu-latest
+    needs: check-commits
+    if: github.event_name == 'pull_request' && needs.check-commits.outputs.matrix != '{}'
+    strategy:
+      fail-fast: false
```

Comment:
why not fail-fast? If some intermediate commit is broken by definition someone needs to fix the commit hence invalidating all results of any commits that come after the broken one?

Reply:
Please undo this change - my understanding was wrong. 

The matrix entries are only created in a specific order - their execution order can be different/concurrent unless `max-parallel: 1` is added.

Each check-commit takes ~10mins at most so doing it serially via `max-parallel: 1` doesn't sound good.

So let's disable `fail-fast` like it was before.

Reply:
This was supposed to be a simple pass/fail check, we don't have to find all broken commits if we already know the check will fail.

Reply:
that's a fair point - we can offload which commit is broken to PR author.